### PR TITLE
feat: multistream-select

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Adds content for Multistream Select and references new content on connections for further explanations.
- Introduces Protocol Select
- addresses #18 

## Latest preview

Not against master (setting this up).

### CODE REVIEW

File: content/concepts/introduction/protocols/multistream.md [hunk 1]
```diff
@@ -1,7 +1,66 @@
...
-description: ""
+description: "Multistream Select is used to negotiate and upgrade the protocol of a connection between two peers in libp2p."
 weight: 26
 ---
 
-Coming soon!
+## Overview
+
+[Multistream Select](https://github.com/multiformats/multistream-select) is
+a protocol for friendly protocol multiplexing that allows for the negotiation
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:10]: What is unfriendly protocol multiplexing? 🙃

File: content/concepts/introduction/protocols/multistream.md [hunk 2]
```diff
@@ -1,7 +1,66 @@
...
+of a [multicodec](https://github.com/multiformats/multicodec) between two entities.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:11]: Is it a multicoded? It's just a string, isn't it?

File: content/concepts/introduction/protocols/multistream.md [hunk 3]
```diff
@@ -1,7 +1,66 @@
...
+> Multicodecs are codecs that can handle multiple different encoding formats.
+> They are used to decode data and are self-describing, meaning they can identify
+> the data format they are decoding. This capability allows nodes to support multiple
+> protocols and efficiently switch between them depending on the task.
+
+Multistream is a general concept and refers to the ability
+to [multiplex multiple streams](../../multiplex/overview.md) of data over a single
+connection. This allows for multiple independent channels of communication to be
+established over a single connection rather than needing to open a separate
+connection for each channel.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:22]: That's multiplexing, not multistream. I don't think we need to explain multiplexing here though.

File: content/concepts/introduction/protocols/multistream.md [hunk 4]
```diff
@@ -1,7 +1,66 @@
...
+
+Multistream Select is a specific protocol built on top of the general
+concept of Multistream. It is used to negotiate and upgrade the protocol of a
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:25]: I'd just call it negotiation.

File: content/concepts/introduction/protocols/multistream.md [hunk 5]
```diff
@@ -1,7 +1,66 @@
...
+connection between two peers and to integrate new protocols as
+they become available. The actual protocol is fairly simple - it is a Multistream
+protocol with a Multistream header.
+
+The header refers to the initial message sent by a peer to initiate the protocol
+negotiation process. This message includes the protocol ID, `/multistream/1.0.0.`,
+which indicates to the listening peer that a protocol negotiation process is about
+to take place. Upon receiving this header, the listening peer responds with their
+desired protocol header.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:34]: I wouldn't put too much emphasis on the header. It's not terribly important. Just say: Every multistream message is prefixed with a header, `/multistream/1.0.0`.

File: content/concepts/introduction/protocols/multistream.md [hunk 6]
```diff
@@ -1,7 +1,66 @@
...
+
+After both peers have sent their headers, both peers need to agree on the version
+of the Multistream Select protocol to use for the connection. Both peers can exchange
+protocol information and select the protocols to use for the connection, including
+protocols for [security](../../secure-comm/overview.md) and
+[stream multiplexing](../../multiplex/overview.md) if necessary. The
+[protocol negotiation](../core-abstractions/connections#protocol-negotiation)
+and [upgrade process](../core-abstractions/connections#upgrading-connections)
+is explained further in the [connections document](../core-abstractions/connections.md).
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:43]: This should be mentioned in the first paragraph:
libp2p uses multistream at various places:
1. a
2. b
3. c

File: content/concepts/introduction/protocols/multistream.md [hunk 7]
```diff
@@ -1,7 +1,66 @@
...
+Protocol Select is in active development and has yet to be fully implemented in
+libp2p. Follow the official specification development
+[here](https://github.com/libp2p/specs/pull/349).
+{{< /alert >}}
+
+> Protocol Select will not be compatible with Multistream Select in its semantics
+> as well as on the wire. Live libp2p-based networks currently using Multistream Select
+> would need to follow a multiphased roll-out strategy to guarantee a smooth transition.
+
+{{< alert icon="💡" context="note" text="See the multistream-select <a class=\"text-muted\" href=\"https://github.com/multiformats/multistream-select\">technical specification</a> for more details." />}}
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:66]: We've actually deprioritized Protocol Select, as the benefits are not really worth it compared to the other items on our roadmap. I suggest removing this paragraph.

I'm currently working on a blog post that will shed some light on the long and windy history of Multistream-Select replacement proposals, and I suggest we link it from here once it's up.

Comment (DannyS03) [reply content/concepts/introduction/protocols/multistream.md:66]: Good to know!

File: content/concepts/introduction/protocols/multistream.md (hunk 1 above)
Comment (DannyS03) [reply content/concepts/introduction/protocols/multistream.md:10]: I suppose multiplexing that does not graceful handle protocol changes, I removed friendly though.

File: content/concepts/introduction/protocols/multistream.md (hunk 5 above)
Comment (DannyS03) [reply content/concepts/introduction/protocols/multistream.md:34]: Simplified

File: content/concepts/introduction/protocols/multistream.md (hunk 6 above)
Comment (DannyS03) [reply content/concepts/introduction/protocols/multistream.md:43]: It should still go after Multistream is introduced.

File: content/concepts/introduction/protocols/multistream.md (hunk 2 above)
Comment (DannyS03) [reply content/concepts/introduction/protocols/multistream.md:11]: Yes.. thanks!

File: content/concepts/introduction/protocols/multistream.md [hunk 8]
```diff
@@ -1,7 +1,23 @@
...
+description: "Multistream Select is used to negotiate the protocol of a connection between two peers in libp2p."
 weight: 26
 ---
 
-Coming soon!
+## Overview
+
+[Multistream Select](https://github.com/multiformats/multistream-select) is a protocol for
+protocol multiplexing that allows for the negotiation of different protocols between two
+entities. It is a specific protocol built on top of the general concept of Multistream.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:11]:
```suggestion
peers.
```

File: content/concepts/introduction/protocols/multistream.md [hunk 9]
```diff
@@ -1,7 +1,23 @@
...
+It is used to negotiate the protocol of a connection between two peers and to integrate
+new protocols as they become available. Every multistream message is prefixed with a
+header, `/multistream/1.0.0`.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:14]:
```suggestion
```

File: content/concepts/introduction/protocols/multistream.md [hunk 10]
```diff
@@ -1,7 +1,23 @@
...
+
+Peers use Multistream Select at various places to select the protocols to use for a
+libp2p connection, including [transports](../../transports/overview.md),
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:17]:
```suggestion
libp2p connection, including 
```

File: content/concepts/introduction/protocols/multistream.md [hunk 11]
```diff
@@ -1,7 +1,23 @@
...
+[security](../../secure-comm/overview.md), and [stream multiplexing](../../multiplex/overview.md)
+as necessary. The [protocol negotiation](../core-abstractions/connections#protocol-negotiation)
+and [upgrade process](../core-abstractions/connections#upgrading-connections) are explained
+further in the [connections document](../core-abstractions/connections.md).
+
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/multistream.md:22]: We should add a sentence saying that it's also used to select application protocols spoken on streams.

Comment (mxinden) [reply content/concepts/introduction/protocols/multistream.md:22]: :+1: done with the last 3 commits.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
@p-shahi good to merge from my side.

**Pull request merged**